packages/cli-clean/src/clean.ts

```diff
@@ -0,0 +1,242 @@
+import {getLoader} from '@react-native-community/cli-tools';
+import type {Config as CLIConfig} from '@react-native-community/cli-types';
+import chalk from 'chalk';
+import execa from 'execa';
+import {existsSync as fileExists, rmdir} from 'fs';
+import os from 'os';
+import path from 'path';
+import prompts from 'prompts';
+import {promisify} from 'util';
+
+type Args = {
+  include?: string;
+  projectRoot: string;
+  verifyCache?: boolean;
+};
+
+type Task = {
+  label: string;
+  action: () => Promise<void>;
+};
+
+type CleanGroups = {
+  [key: string]: {
+    description: string;
+    tasks: Task[];
+  };
+};
+
+const DEFAULT_GROUPS = ['metro', 'watchman'];
+
+const rmdirAsync = promisify(rmdir);
+
+function cleanDir(directory: string): Promise<void> {
+  if (!fileExists(directory)) {
+    return Promise.resolve();
+  }
+
+  return rmdirAsync(directory, {maxRetries: 3, recursive: true});
+}
+
+function findPath(startPath: string, files: string[]): string | undefined {
+  // TODO: Find project files via `@react-native-community/cli`
+  for (const file of files) {
+    const filename = path.resolve(startPath, file);
+    if (fileExists(filename)) {
+      return filename;
+    }
+  }
+
+  return undefined;
+}
+
+async function promptForCaches(
+  groups: CleanGroups,
+): Promise<string[] | undefined> {
+  const {caches} = await prompts({
+    type: 'multiselect',
+    name: 'caches',
+    message: 'Select all caches to clean',
+    choices: Object.entries(groups).map(([cmd, group]) => ({
+      title: `${cmd} ${chalk.dim(`(${group.description})`)}`,
+      value: cmd,
+      selected: DEFAULT_GROUPS.includes(cmd),
+    })),
+    min: 1,
+  });
+  return caches;
+}
+
+export async function clean(
+  _argv: string[],
+  _config: CLIConfig,
+  cleanOptions: Args,
+): Promise<void> {
+  const {include, projectRoot, verifyCache} = cleanOptions;
+  if (!fileExists(projectRoot)) {
+    throw new Error(`Invalid path provided! ${projectRoot}`);
+  }
+
+  const COMMANDS: CleanGroups = {
+    android: {
+      description: 'Android build caches, e.g. Gradle',
+      tasks: [
+        {
+          label: 'Clean Gradle cache',
+          action: async () => {
+            const candidates =
+              os.platform() === 'win32'
+                ? ['android/gradlew.bat', 'gradlew.bat']
+                : ['android/gradlew', 'gradlew'];
+            const gradlew = findPath(projectRoot, candidates);
+            if (gradlew) {
+              const script = path.basename(gradlew);
+              await execa(
+                os.platform() === 'win32' ? script : `./${script}`,
+                ['clean'],
```

Comment:
Small FYI: clean won't work for users on New Architecture due to a bug on AGP that is causing `clean` to depend on `preBuild`. I agree that `./gradlew clean` would be the correct approach here, but maybe we should not advertise it yet? Or have a `isNewArchitectureEnabled` flag inside the CLI to check this and run a `rm`?